<html>
<head>
<meta http-equiv="Content-Type" content="text/html;charset=gbk">
<title>PostgreSQL research</title>
<style type="text/css">
.blogcnt{line-height:160%;font-size:14px;text-align:left;word-wrap:break-word;}
.blogcnt *{line-height:160%;}
.blogcnt p{margin:0 0 10px;}
.blogcnt ul,.nbw-blog ol{margin:5px 0 5px 40px;padding:0}
.blogcnt em{font-style:italic;}
.blogcnt blockquote{font-size:1em;margin:auto 0 auto 35px;}
.blogcnt img{border:0;max-width:100%;}
</style>
</head>
<body style="color:#444444;">
<h1 id="blog-Title"><a href="index.html">PostgreSQL research</a></h1>
<div id="" style="padding:0 20px;">
	<h2 id="">Lua error message and traceback</h2>
	<h5 id="">2014-02-20 9:56:01&nbsp;&nbsp;&nbsp;<a href="http://blog.163.com/digoal@126/blog/static/163877040201412093914335/" target="_blank">查看原文&gt;&gt;</a></h5>
	<div class="" id="" style="padding:0 20px;">
		<div class="blogcnt" style="width:800px;"><div>Lua执行遇到错误时, 如果错误信息输出为字符串类型, 那么还会附加两个内容,&nbsp;</div><div>1. 输出来源, 例如文件或标准输入</div><div>2. 错误位置, 行号.&nbsp;</div><div>例如 :&nbsp;</div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; function foo(str)</font></div><div><font size="2"   >&gt;&gt; &nbsp; if type(str) ~= 'string' then</font></div><div><font size="2"   >&gt;&gt; &nbsp; &nbsp; print("line 1\n")&nbsp;</font></div><div><font size="2"   >&gt;&gt; &nbsp; &nbsp; error("string expected")</font></div><div><font size="2"   >&gt;&gt; &nbsp; end</font></div><div><font size="2"   >&gt;&gt; &nbsp; print("end")</font></div><div><font size="2"   >&gt;&gt; end</font></div><div><font size="2"   >&gt; foo(1)</font></div><div><font size="2"   >line 1</font></div><div><font size="2"   ><br></font></div><div><font size="2"   >stdin:4: string expected &nbsp;-- 这行是错误信息, 包括错误信息的来源stdin, 以及行号4. foo函数第四行报错.</font></div><div><font size="2"   >stack traceback: &nbsp;-- 这里开始时traceback信息.</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'error'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:4: in function 'foo'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</font></div><p></p></pre></div><div><br></div><div>如果把error的第二个参数(<span style="line-height: 28px;"   >默认是1</span><span style="line-height: 28px;"   >)改成2, 那么输出的行号会变成1, 实际触发错误的位置是error函数, 2表示层级.&nbsp;</span></div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; function foo(str)</font></div><div><font size="2"   >&nbsp; if type(str) ~= 'string' then</font></div><div><font size="2"   >&nbsp; &nbsp; print("line 1\n")&nbsp;</font></div><div><font size="2"   >&nbsp; &nbsp; error("string expected", 2)</font></div><div><font size="2"   >&nbsp; end</font></div><div><font size="2"   >&nbsp; print("end")</font></div><div><font size="2"   >end</font></div><div><font size="2"   >&gt; foo(1)</font></div><div><font size="2"   >line 1</font></div><div><font size="2"   ><br></font></div><div><font size="2"   >stdin:1: string expected</font></div><div><font size="2"   >stack traceback:</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'error'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:4: in function 'foo'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</font></div><p></p></pre></div><div><br></div><div>traceback的信息</div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; do</font></div><div><font size="2"   >print("hello")</font></div><div><font size="2"   >foo(1)</font></div><div><font size="2"   >end</font></div><div><font size="2"   >hello</font></div><div><font size="2"   >line 1</font></div><div><font size="2"   ><br></font></div><div><font size="2"   >stdin:4: string expected</font></div><div><font size="2"   >stack traceback:</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'error'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:4: in function 'foo'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:3: in main chunk &nbsp;-- 如果触发错误的函数在main chunk中第三行, 则这里输出第三行.</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</font></div><p></p></pre></div><div><br></div><div>注意, 使用pcall的话, msg将缺失traceback的信息. 使用xpcall解决.</div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >函数 :&nbsp;</font></div><div><div><font size="2"   >&gt; function foo(str)</font></div><div><font size="2"   >&nbsp; if type(str) ~= 'string' then</font></div><div><font size="2"   >&nbsp; &nbsp; print("line 1\n")&nbsp;</font></div><div><font size="2"   >&nbsp; &nbsp; error("string expected")</font></div><div><font size="2"   >&nbsp; end</font></div><div><font size="2"   >&nbsp; print("end")</font></div><div><font size="2"   >end</font></div></div><div><font size="2"   >使用pcall调用foo, 传入参数1.</font></div><div><div><font size="2"   >&gt; res, msg = pcall(foo, 1)</font></div><div><font size="2"   >line 1</font></div><div><font size="2"   ><br></font></div><div><font size="2"   >&gt; print(res)</font></div><div><font size="2"   >false</font></div><div><font size="2"   >&gt; print(msg) &nbsp;-- msg中没有存储traceback的信息.</font></div><div><font size="2"   >stdin:4: string expected</font></div></div><div><font size="2"   ><br></font></div><div><font size="2"   >使用xpcall可以解决这个问题, 比pcall多了一个参数, error handler function. 这里使用debug.traceback.</font></div><div><div><font size="2"   >&gt; res, msg = xpcall(foo, debug.traceback, 1)</font></div><div><font size="2"   >line 1</font></div><div><font size="2"   ><br></font></div><div><font size="2"   >&gt; print(msg) -- 使用debug.traceback可以将traceback的信息存储到msg变量.</font></div><div><font size="2"   >stdin:4: string expected</font></div><div><font size="2"   >stack traceback:</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'error'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:4: in function &lt;stdin:1&gt;</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in function 'xpcall'</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; stdin:1: in main chunk</font></div><div><font size="2"   >&nbsp; &nbsp; &nbsp; &nbsp; [C]: in ?</font></div><div><font size="2"   >&gt; print(res)</font></div><div><font size="2"   >false</font></div></div><p></p></pre></div><div><br></div><div>如果使用debug.debug作为error handler的话, 那么将进入handler命令行.</div><div><pre class="prettyprint"   ><p></p><div><font size="2"   >&gt; res, msg = xpcall(foo, debug.debug, 1)</font></div><div><font size="2"   >line 1</font></div><div><font size="2"   ><br></font></div><div><font size="2"   >lua_debug&gt; 这里可以执行一些东西</font></div><p></p></pre></div><div><br></div>[参考]<wbr><div>1.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-error" rel="nofollow"   >error (message [, level])</a></h3><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Terminates the last protected function called and returns&nbsp;</span><code style="font-size: medium; line-height: normal; text-align: justify;"   >message</code><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;as the error message. Function&nbsp;</span><code style="font-size: medium; line-height: normal; text-align: justify;"   >error</code><span style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >&nbsp;never returns.</span><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Usually,&nbsp;<code style="font-size: 12pt;"   >error</code>&nbsp;adds some information about the error position at the beginning of the message, if the message is a string. The&nbsp;<code style="font-size: 12pt;"   >level</code>&nbsp;argument specifies how to get the error position. With level&nbsp;1 (the default), the error position is where the&nbsp;<code style="font-size: 12pt;"   >error</code>&nbsp;function was called. Level&nbsp;2 points the error to where the function that called&nbsp;<code style="font-size: 12pt;"   >error</code>&nbsp;was called; and so on. Passing a level&nbsp;0 avoids the addition of error position information to the message.</p></div><div>2.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-pcall" rel="nofollow"   >pcall (f [, arg1, ・・・])</a></h3><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Calls function&nbsp;<code style="font-size: 12pt;"   >f</code>&nbsp;with the given arguments in&nbsp;<em>protected mode</em>. This means that any error inside&nbsp;<code style="font-size: 12pt;"   >f</code>&nbsp;is not propagated; instead,&nbsp;<code style="font-size: 12pt;"   >pcall</code>&nbsp;catches the error and returns a status code. Its first result is the status code (a boolean), which is true if the call succeeds without errors. In such case,&nbsp;<code style="font-size: 12pt;"   >pcall</code>&nbsp;also returns all results from the call, after this first result. In case of any error,<code style="font-size: 12pt;"   >pcall</code>&nbsp;returns&nbsp;<b>false</b>&nbsp;plus the error message.</p></div><div>3.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-xpcall" rel="nofollow"   >xpcall (f, msgh [, arg1, ・・・])</a></h3><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >This function is similar to&nbsp;<a style="color: rgb(0, 0, 128); background-color: inherit; text-decoration: none;" rel="nofollow" href="http://www.lua.org/manual/5.2/manual.html#pdf-pcall"   ><code style="font-size: 12pt;"   >pcall</code></a>, except that it sets a new message handler&nbsp;<code style="font-size: 12pt;"   >msgh</code>.</p></div><div>4.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-debug.debug" rel="nofollow"   >debug.debug ()</a></h3><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Enters an interactive mode with the user, running each string that the user enters. Using simple commands and other debug facilities, the user can inspect global and local variables, change their values, evaluate expressions, and so on. A line containing only the word&nbsp;<code style="font-size: 12pt;"   >cont</code>&nbsp;finishes this function, so that the caller continues its execution.</p><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >Note that commands for&nbsp;<code style="font-size: 12pt;"   >debug.debug</code>&nbsp;are not lexically nested within any function and so have no direct access to local variables.</p></div><div>5.&nbsp;</div><div><h3 style="font-family: Verdana, Geneva, sans-serif; font-weight: normal; padding-left: 0.5em; border-left-style: solid; border-left-color: rgb(208, 208, 255); border-left-width: 1em; line-height: normal; text-align: justify;"   ><a name="pdf-debug.traceback" rel="nofollow"   >debug.traceback ([thread,] [message [, level]])</a></h3><p style="font-family: Helvetica, Arial, sans-serif; font-size: medium; line-height: normal; text-align: justify;"   >If&nbsp;<code style="font-size: 12pt;"   >message</code>&nbsp;is present but is neither a string nor&nbsp;<b>nil</b>, this function returns&nbsp;<code style="font-size: 12pt;"   >message</code>&nbsp;without further processing. Otherwise, it returns a string with a traceback of the call stack. An optional<code style="font-size: 12pt;"   >message</code>&nbsp;string is appended at the beginning of the traceback. An optional&nbsp;<code style="font-size: 12pt;"   >level</code>&nbsp;number tells at which level to start the traceback (default is 1, the function calling&nbsp;<code style="font-size: 12pt;"   >traceback</code>).</p></div></div>
	</div>
</div>
</body>
</html>